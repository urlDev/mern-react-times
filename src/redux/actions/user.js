import axios from 'axios';
import toaster from 'toasted-notes';
import { push } from 'connected-react-router';

import NotificationComponent from '../../components/notification-component/NotificationComponent';

const url = process.env.REACT_APP_API_URL;

export const REGISTER_USER = 'REGISTER_USER';
export const LOGOUT_USER = 'LOGOUT_USER';
export const LOGIN_USER = 'LOGIN_USER';
export const DELETE_USER = 'DELETE_USER';
export const UPDATE_USER = 'UPDATE_USER';
export const GET_USER = 'GET_USER';
export const SET_TOKEN = 'SET_TOKEN';
export const USER_FETCH_ERROR = 'USER_FETCH_ERROR';
export const USER_LOADING = 'USER_LOADING';
export const USER_LOADING_END = 'USER_LOADING_END';
export const USER_MODAL_OPEN = 'USER_MODAL_OPEN';
export const USER_MODAL_CLOSE = 'USER_MODAL_CLOSE';
export const DELETE_MODAL_OPEN = 'DELETE_MODAL_OPEN';
export const DELETE_MODAL_CLOSE = 'DELETE_MODAL_CLOSE';
export const UPLOAD_AVATAR = 'UPLOAD_AVATAR';

export const registerUser = (user) => ({
  type: REGISTER_USER,
  payload: user,
});

export const logOutUser = () => ({
  type: LOGOUT_USER,
});

export const loginUser = (user) => ({
  type: LOGIN_USER,
  payload: user,
});

export const deleteUser = () => ({
  type: DELETE_USER,
});

export const updateUser = (user) => ({
  type: UPDATE_USER,
  payload: user,
});

export const getUser = (user) => ({
  type: GET_USER,
  payload: user,
});

export const setToken = (token) => ({
  type: SET_TOKEN,
  payload: token,
});

export const userFetchError = (error) => ({
  type: USER_FETCH_ERROR,
  payload: error,
});

export const userLoading = () => ({
  type: USER_LOADING,
});

export const userLoadingEnd = () => ({
  type: USER_LOADING_END,
});

export const userModalOpen = () => ({
  type: USER_MODAL_OPEN,
});

export const userModalClose = () => ({
  type: USER_MODAL_CLOSE,
});

export const deleteModalOpen = () => ({
  type: DELETE_MODAL_OPEN,
});

export const deleteModalClose = () => ({
  type: DELETE_MODAL_CLOSE,
});

export const uploadAvatar = (images) => ({
  type: UPLOAD_AVATAR,
  payload: images,
});

export const fetchRegisterUser = (user) => async (dispatch) => {
  dispatch(userLoading());
  try {
    const response = await axios.post(`${url}/profile/register`, user);
    const data = await response.data;
    return [
      dispatch(registerUser(data.user)),
      dispatch(userLoadingEnd()),
      dispatch(setToken(data.token)),
      localStorage.setItem('token', JSON.stringify(data.token)),
      localStorage.setItem('user', JSON.stringify(data.user)),
      dispatch(push('/')),
      toaster.notify(
        () => (
          <NotificationComponent
            success={true}
            text={`Welcome! Enjoy your time`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      dispatch(userLoadingEnd()),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Oops! Something went wrong!'}
            success={false}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};

export const fetchLogoutUser = () => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };

  const path = window.location.pathname;

  try {
    await axios.post(`${url}/profile/logout`, null, config);

    return [
      path.includes('profile') && dispatch(push('/home')),
      dispatch(logOutUser()),
      toaster.notify(
        () => <NotificationComponent text={'Buh-Bye!'} success={true} />,
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return dispatch(userFetchError(error));
  }
};

export const fetchLoginUser = (user) => async (dispatch) => {
  dispatch(userLoading());
  try {
    const response = await axios.post(`${url}/profile/login`, user);
    const data = await response.data;
    return [
      dispatch(loginUser(data.user)),
      dispatch(userLoadingEnd()),
      dispatch(setToken(data.token)),
      localStorage.setItem('token', JSON.stringify(data.token)),
      localStorage.setItem('user', JSON.stringify(data.user)),
      dispatch(push('/home')),
      toaster.notify(
        () => (
          <NotificationComponent
            success={true}
            text={`Welcome! Enjoy your time`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      dispatch(userLoadingEnd()),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Oops! Something went wrong!'}
            success={false}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};

// To get the user,
// but because I added an or statement in user and token initial states,
// app will check if there is user and token in localStorage
// so there will be no need for this.
// Adding just in case.
export const fetchUser = () => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };
  dispatch(userLoading());
  try {
    const response = await axios.get(`${url}/profile`, config);
    const data = await response.data;
    return [
      dispatch(registerUser(data.user)),
      dispatch(userLoadingEnd()),
      dispatch(setToken(data.token)),
      localStorage.setItem('token', JSON.stringify(data.token)),
      localStorage.setItem('user', JSON.stringify(data.user)),
    ];
  } catch (error) {
    return [dispatch(userFetchError(error)), dispatch(userLoadingEnd())];
  }
};

export const fetchUpdateUser = (user) => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };
  try {
    const response = await axios.patch(`${url}/profile`, user, config);
    const data = await response.data;
    return [
      dispatch(updateUser(data)),
      localStorage.setItem('user', JSON.stringify(data)),
      dispatch(push('/')),
      toaster.notify(
        () => (
          <NotificationComponent
            success={true}
            text={`Updated successfully!`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Oops! Something went wrong!'}
            success={false}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};

export const fetchDeleteUser = () => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };
  try {
    await axios.delete(`${url}/profile`, config);
    return [
      localStorage.clear(),
      dispatch(push('/')),
      dispatch(deleteUser()),
      toaster.notify(
        () => <NotificationComponent success={true} text={`Buh-bye!`} />,
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      toaster.notify(
        () => (
          <NotificationComponent
            success={false}
            text={`Something went wrong!`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};

export const fetchUploadAvatar = (input) => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };

  try {
    const response = await axios.post(`${url}/profile/avatar`, input, config);
    const data = await response.data;
    return [
      dispatch(uploadAvatar(data)),
      localStorage.setItem('user', JSON.stringify(data)),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Avatar is changed! Wow, that looks amazing! ;)'}
            success={true}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Oops! Something went wrong!'}
            success={false}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};
