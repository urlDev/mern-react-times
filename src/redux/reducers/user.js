import {
    REGISTER_USER,
    SET_TOKEN,
    USER_FETCH_ERROR,
    USER_LOADING,
    USER_LOADING_END,
} from '../actions/user';

const initialState = {
    loadingUser: false,
    errorUser: null,
    user: JSON.parse(localStorage.getItem('user')) || {},
    token: JSON.parse(localStorage.getItem('token')) || {},
};

const userReducer = (state = initialState, action) => {
    switch (action.payload) {
        case REGISTER_USER:
            return {
                ...state,
                user: action.payload,
            };
        case SET_TOKEN:
            return {
                ...state,
                token: action.payload,
            };
        case USER_FETCH_ERROR:
            return {
                ...state,
                error: action.payload,
            };
        case USER_LOADING:
            return {
                ...state,
                loading: true,
            };
        case USER_LOADING_END:
            return {
                ...state,
                loading: false,
            };
        default:
            return state;
    }
};

export default userReducer;