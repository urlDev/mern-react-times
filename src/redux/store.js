import { createStore, applyMiddleware } from 'redux';
import rootReducer from './rootReducer';

import thunk from 'redux-thunk';
import logger from 'redux-logger';

const middlewares = [thunk, logger];

export default createStore(rootReducer, applyMiddleware(...middlewares));