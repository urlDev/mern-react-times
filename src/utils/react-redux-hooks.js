// Mocking redux hooks
import {
  useSelector as originalUseSelector,
  useDispatch as originalUseDispatch,
} from "react-redux";

export const useSelector = (state) => originalUseSelector(state);
export const useDispatch = () => originalUseDispatch();
