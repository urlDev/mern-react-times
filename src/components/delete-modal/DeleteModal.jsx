import React from "react";
import { useDispatch } from "utils/react-redux-hooks";

import { deleteModalClose, fetchDeleteUser } from "redux/actions/user";
import {
  DeleteButtonContainer,
  DeleteModalContainer,
  DeleteTextAndButtons,
  ModalContainerAllPage,
} from "./DeleteModal.styles";

const DeleteModal = () => {
  const [scroll, setScroll] = React.useState(window.scrollY);
  const dispatch = useDispatch();

  const handleClick = () => {
    dispatch(fetchDeleteUser());
    dispatch(deleteModalClose());
  };

  const updateScroll = () => {
    setScroll(window.scrollY);
  };

  // When user views the delete modal in small screens,
  // there was a bug related to where the user views it.
  // Because the top value was 0, it always put the modal to top.
  // And even though bottom was 0 too, modal opened regarding to viewport height.
  // So, by adding scroll, I could make the modal open where the user is.
  // This is the js solution.
  // In CSS, I can make it happen with only position:fixed
  // and voila!
  React.useEffect(() => {
    document.addEventListener("scroll", updateScroll);
    updateScroll();
    return () => {
      document.removeEventListener("scroll", updateScroll);
    };
  }, []);

  return (
    <ModalContainerAllPage onClick={() => dispatch(deleteModalClose())}>
      <DeleteModalContainer>
        <DeleteTextAndButtons>
          <h1>Sure about this?</h1>
          <DeleteButtonContainer>
            <button onClick={() => dispatch(deleteModalClose())}>Cancel</button>
            <button onClick={() => handleClick()}>Delete</button>
          </DeleteButtonContainer>
        </DeleteTextAndButtons>
      </DeleteModalContainer>
    </ModalContainerAllPage>
  );
};

export default DeleteModal;
