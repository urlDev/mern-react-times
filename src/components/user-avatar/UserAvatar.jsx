import React from "react";
import { useDispatch } from "utils/react-redux-hooks";

import { fetchUploadAvatar } from "redux/actions/user";

import UserImage from "components/user-image/UserImage";

import { ImageAndButton } from "./UserAvatar.styles";

const UserAvatar = () => {
  const dispatch = useDispatch();

  const handleChange = async (e) => {
    const data = new FormData();

    // Naming it avatar because API expects that name
    data.append("avatar", e.target.files[0]);
    await dispatch(fetchUploadAvatar(data));
  };

  return (
    <ImageAndButton>
      <UserImage width="100%" />
      <form>
        <label htmlFor="single">Change Avatar*</label>
        <input type="file" id="single" onChange={handleChange} />
      </form>
      <h2>
        *Should be smaller than <span>2MB</span>
      </h2>
    </ImageAndButton>
  );
};

export default UserAvatar;
