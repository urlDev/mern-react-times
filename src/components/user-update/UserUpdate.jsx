import React from "react";
import { useDispatch, useSelector } from "utils/react-redux-hooks";

import { fetchUpdateUser } from "redux/actions/user";

import { UserDetailsForm } from "./UserUpdate.styles";

const UserUpdate = () => {
  const { user } = useSelector((state) => state.user);
  // To be able to use user data as default, I need to pass them in first
  // This way I can use them as default values in form before updating
  const [{ name, email, password }, setInput] = React.useState({
    name: user.name,
    email: user.email,
  });

  const dispatch = useDispatch();

  const handleChange = (e) =>
    setInput((prevState) => ({
      ...prevState,
      [e.target.name]: e.target.value,
    }));

  const handleSubmit = async (e) => {
    e.preventDefault();
    await dispatch(fetchUpdateUser({ name, email, password }));
  };

  return (
    <UserDetailsForm>
      <form onSubmit={handleSubmit}>
        <div>
          <label htmlFor="name">Name / Username</label>
          <input type="text" name="name" value={name} onChange={handleChange} />
        </div>
        <div>
          <label htmlFor="email">Email</label>
          <input
            type="email"
            name="email"
            value={email}
            onChange={handleChange}
          />
        </div>
        <div>
          <label htmlFor="password">Password</label>
          <input
            type="password"
            name="password"
            autoComplete="off"
            onChange={handleChange}
            required
          />
        </div>
        <div>
          <button aria-label="save user details" type="submit">
            Update
          </button>
        </div>
      </form>
    </UserDetailsForm>
  );
};

export default UserUpdate;
