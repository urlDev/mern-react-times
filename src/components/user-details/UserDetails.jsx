import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useHistory } from 'react-router';

import { deleteModalOpen, fetchUpdateUser } from '../../redux/actions/user';

import ErrorFallback from '../error-fallback/ErrorFallback';
import MarketComponents from '../market-components/MarketComponents';
import Nav from '../nav/Nav';
import StoryTopicHeaders from '../story-topic-headers/StoryTopicHeaders';
import UserImage from '../user-image/UserImage';
import DeleteModal from '../delete-modal/DeleteModal';
import LogoDarkSrc from '../../assets/logo.svg';

import { StoryTopicContainer } from '../story-topic/StoryTopic.styles.js';
import {
  UserDetailContainer,
  ImageAndButton,
  UserDetailsForm,
  DeleteAccount,
} from './UserDetails.styles';

const UserDetails = () => {
  const [{ name, email, password }, setInput] = React.useState({});
  const { errorChart } = useSelector((chart) => chart.chart);
  const { user, deleteModal } = useSelector((user) => user.user);
  const dispatch = useDispatch();
  const history = useHistory();

  const handleChange = (e) =>
    setInput((prevState) => ({
      ...prevState,
      [e.target.name]: e.target.value,
    }));

  const handleSubmit = async (e) => {
    e.preventDefault();
    await dispatch(fetchUpdateUser({ name, email, password }));
    history.push('/');
  };

  return (
    <>
      <Nav logo={LogoDarkSrc} borderBottom="1px solid lightgray" icon="1" />
      <StoryTopicHeaders />
      <div style={{ margin: '0 30px' }}>
        <StoryTopicContainer>
          <h1 style={{ marginTop: '25px' }}>Profile</h1>
        </StoryTopicContainer>

        <UserDetailContainer>
          <ImageAndButton>
            <UserImage width="100%" />
            <h2>
              *Should be smaller than <span>2MB</span>
            </h2>
            <button>Change avatar</button>
          </ImageAndButton>
          <UserDetailsForm>
            <form onSubmit={handleSubmit}>
              <div>
                <label htmlFor="name">Name / Username</label>
                <input
                  type="text"
                  name="name"
                  defaultValue={user.name}
                  value={name}
                  onChange={handleChange}
                />
              </div>
              <div>
                <label htmlFor="email">Email</label>
                <input
                  type="email"
                  name="email"
                  defaultValue={user.email}
                  value={email}
                  onChange={handleChange}
                />
              </div>
              <div>
                <label htmlFor="password">Password</label>
                <input
                  type="password"
                  name="password"
                  autoComplete="off"
                  onChange={handleChange}
                  required
                />
              </div>
              <div>
                <button aria-label="save user details" type="submit">
                  Update
                </button>
              </div>
            </form>
          </UserDetailsForm>
          <DeleteAccount>
            <div>
              <h1>I want to delete my account!</h1>
              <button onClick={() => dispatch(deleteModalOpen())}>
                Delete
              </button>
            </div>
          </DeleteAccount>
        </UserDetailContainer>
        {deleteModal ? <DeleteModal /> : null}
      </div>
      {errorChart ? <ErrorFallback /> : <MarketComponents />}
    </> //
  );
};

export default UserDetails;
