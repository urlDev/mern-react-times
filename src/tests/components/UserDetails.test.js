import { shallow } from "enzyme";

import * as ReactReduxHooks from "utils/react-redux-hooks";
import { mockStore } from "tests/store";

import { user } from "tests/fixtures/user";

import UserDetails from "components/user-details/UserDetails";
import UserAvatar from "components/user-avatar/UserAvatar";
import UserUpdate from "components/user-update/UserUpdate";
import DeleteModal from "components/delete-modal/DeleteModal";

describe("Testing UserDetails component", () => {
  let wrapper;
  let store;

  beforeEach(() => {
    jest
      .spyOn(ReactReduxHooks, "useSelector")
      .mockImplementation((state) => store.getState());
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
  });

  describe("Testing the component with user", () => {
    beforeEach(() => {
      store = mockStore({
        user,
        deleteModal: false,
        width: 1100,
      });

      wrapper = shallow(<UserDetails store={store} />);
    });

    test("Should render the component when width is smaller than 1150", () => {
      expect(wrapper).toHaveLength(1);
      expect(wrapper.containsMatchingElement(<UserUpdate />)).toBeTruthy();
    });
    test("Should match snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });
    test("Should render the component when width is bigger than 1150", () => {
      store = mockStore({
        width: 1200,
        user,
        deleteModal: false,
      });

      wrapper = shallow(<UserDetails store={store} />);
      expect(wrapper).toHaveLength(1);
      expect(wrapper).toMatchSnapshot();
    });
  });
  describe("Testing component without user", () => {
    beforeEach(() => {
      store = mockStore({
        user: {},
        width: 1200,
        deleteModal: false,
      });

      wrapper = shallow(<UserDetails store={store} />);
    });

    test("Should render component without user", () => {
      expect(wrapper).toHaveLength(1);
      expect(wrapper.containsMatchingElement(<UserAvatar />)).toBeFalsy();
      expect(wrapper.containsMatchingElement(<DeleteModal />)).toBeFalsy();
    });
    test("Should match snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });
    test("Should show delete modal", () => {
      store = mockStore({
        user,
        deleteModal: true,
        width: 1100,
      });

      wrapper = shallow(<UserDetails />);

      expect(wrapper).toHaveLength(1);
      expect(wrapper.containsMatchingElement(<DeleteModal />)).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
