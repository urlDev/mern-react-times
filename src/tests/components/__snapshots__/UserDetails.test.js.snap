// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing UserDetails component Testing component without user Should match snapshot 1`] = `
<Fragment>
  <ResponsiveNav />
  <div
    style={
      Object {
        "margin": "0 15px",
      }
    }
  >
    <styled.div>
      <h1
        style={
          Object {
            "marginTop": "25px",
          }
        }
      >
        Profile
      </h1>
    </styled.div>
  </div>
  <ErrorBoundary
    FallbackComponent={[Function]}
  >
    <MarketComponents />
  </ErrorBoundary>
</Fragment>
`;

exports[`Testing UserDetails component Testing component without user Should show delete modal 1`] = `
<Fragment>
  <ResponsiveNav />
  <div
    style={
      Object {
        "margin": "0 15px",
      }
    }
  >
    <styled.div>
      <h1
        style={
          Object {
            "marginTop": "25px",
          }
        }
      >
        Profile
      </h1>
    </styled.div>
    <styled.div>
      <UserAvatar />
      <styled.div>
        <UserUpdate />
        <UserAccountDelete />
      </styled.div>
    </styled.div>
    <DeleteModal />
  </div>
  <ErrorBoundary
    FallbackComponent={[Function]}
  >
    <MarketComponents />
  </ErrorBoundary>
</Fragment>
`;

exports[`Testing UserDetails component Testing the component with user Should match snapshot 1`] = `
<Fragment>
  <ResponsiveNav />
  <div
    style={
      Object {
        "margin": "0 15px",
      }
    }
  >
    <styled.div>
      <h1
        style={
          Object {
            "marginTop": "25px",
          }
        }
      >
        Profile
      </h1>
    </styled.div>
    <styled.div>
      <UserAvatar />
      <styled.div>
        <UserUpdate />
        <UserAccountDelete />
      </styled.div>
    </styled.div>
  </div>
  <ErrorBoundary
    FallbackComponent={[Function]}
  >
    <MarketComponents />
  </ErrorBoundary>
</Fragment>
`;

exports[`Testing UserDetails component Testing the component with user Should render the component when width is bigger than 1150 1`] = `
<Fragment>
  <ResponsiveNav />
  <div
    style={
      Object {
        "margin": "0 15px",
      }
    }
  >
    <styled.div>
      <h1
        style={
          Object {
            "marginTop": "25px",
          }
        }
      >
        Profile
      </h1>
    </styled.div>
    <styled.div>
      <UserAvatar />
      <UserUpdate />
      <UserAccountDelete />
    </styled.div>
  </div>
  <ErrorBoundary
    FallbackComponent={[Function]}
  >
    <MarketComponents />
  </ErrorBoundary>
</Fragment>
`;
